### ISSUE TITLE
Lending rate error

### ISSUE THREAD

**Issue by rreceptai** [Mikadily/poloniexlendingbot#26]
Bot keeps on lending my ETH at 4.99%, but the gap top and gap bottom is set from 1% to 30%
Lending btc works - ok.
I hope this can be easily solved.

![lending-rate-error](https://cloud.githubusercontent.com/assets/7442312/13601063/f48f0a06-e535-11e5-8101-370407baa145.jpg)

I have the following config active:
```
[BOT]
#sleep between iterations, time in seconds
sleeptime = 10
#minimum daily lend rate in percent
mindailyrate = 0.05
#max rate. 2% is good choice because it's default at margin trader interface. 5% is max to be accepted by the exchange
maxdailyrate = 5
#The number of offers to split the available balance uniformly across the [gaptop, gapbottom] range.
spreadlend = 10
#The depth of lendbook (in percent of lendable balance) to move through before placing the first (gapbottom) and last (gaptop) offer.
#if gapbottom is set to 0, the first offer will be at the lowest possible rate. However some low value is recommended (say 10%) to skip dust offers
gapbottom = 1
gaptop = 30
#Daily lend rate threshold after which we offer lends for 60 days as opposed to 2. If set to 0 all offers will be placed for a 2 day period
sixtydaythreshold = 0.5
# AutoRenew - if set to 1 the bot will toggle the AutoRenew flag for the loans when you stop it (Ctrl+C) and clear the AutoRenew flag when started
autorenew = 0
#custom config per coin, useful when closing positions etc.
#syntax: [COIN:mindailyrate:maxactiveamount, ... COIN:mindailyrate:maxactiveamount]
#if maxactive amount is 0 - stop lending this coin. in the future you'll be able to limit amount to be lent.
#coinconfig = ["BTC:0.05:1"]
```

**Comment by rnevet** [comment]
mindailyrate = 0.05
means you won't lend for less then 0.05% ~ 0.0499% that you are getting for ETH.
Use the coinconfig option to create different configs for BTC and ETH.

**Comment by rreceptai** [comment]
I understand how to use mindailyrate, the bot offers loans at 4.99%, which is not normal - check the screenshot and my configuration.

**Comment by rnevet** [comment]
Sorry about that, didn't notice the open loan... that is weird.

**Comment by rnevet** [comment]
This is still the result of the calculated loan rate being lower then the minDailyRate, the fallback is to create loans with maxDailyRate (maybe it shouldn't).
To solve this you can use the "coinconfig", just uncomment and add the ETH config:
coinconfig = ["BTC:0.09:5","ETH:0.03:5"]

**Comment by rreceptai** [comment]
According to my configuration it should place loans from 0.05% to 5% i do not see why it should start with the full amount at 4.99%

**Comment by rnevet** [comment]
I agree that it shouldn't loan at all, right now the behavior is to create a loan at maxRate.

**Comment by Mikadily** [comment]
The order book is out of bounds of your minlendingrate. I assume at the time of writing the range you could see the order book was something 0.0367-0.4555%, therefore bot assumes there's nothing to compete against, and go straight to the max. It will correct it self once it can see more orders.
It's the correct behavior, it's the way I intended it to be. And I know it confuses many people. Maybe a warning message in the logs would help.

**Comment by Mikadily** [comment]
Oh. Now I noticed, on the Poloniex website you can 'Load more' Loans. Before you couldn't. The bot only sees top-20 (or something) lowest loans. Will see how we can improve it.

**Comment by rnevet** [comment]
It's not in the API for returnLoanOrders, but the website is using a limit parameter to limit the amount of records.
https://poloniex.com/public?command=returnLoanOrders&currency=ETH&**limit=5**  will return only 5.
So using a high number 9999 will return all of them.

On the other side it will make the bot work harder? ;)

**Comment by dutu** [comment]
@Mikadily, API call `returnLoanOrders` accepts parameter `limit` (which is not documented).
Setting the `limit `to a high value (e.g. 1000), will give you the entire lend-book.

**Comment by Mikadily** [comment]
@rnevet personally I'm happy with how it is at the moment. Maybe it could be a config option - select your limit. And a warning when it happens, to inform the confused user.

@dutu thanks. It's new to me.

**Comment by rnevet** [comment]
@Mikadily I'm also content with they way it's working.
I'm not sure in what improvement would come from increasing the amount of offers retrieved from the server.  For me it would be a disadvantage as I'm running it on my Android phone and this will increase my data usage. 

I would consider that if the offers are lower than minDailyRate the new offer created would be at minDailyRate and not at maxDailyRate? just a thought. So for example in this case the offer created would be at 0.05%.   Maybe only offer a 1/3 of the available amount. 

@rreceptai What did you actually expect?

**Comment by Mikadily** [comment]
@rnevet 
> I would consider that if the offers are lower than minDailyRate the new offer created would be at minDailyRate and not at maxDailyRate? just a thought. So for example in this case the offer created would be at 0.05%. Maybe only offer a 1/3 of the available amount.

No. Imagine a legitimate loan book:
0.01%
0.02%
0.03%

Just 3 loan offer on the book and my mindailyrate at 0.05%. I don't want my order set at 0.05%. I want it at max percentage once the other offers are depleted.

**Comment by rnevet** [comment]
I see, it's a different case if the book is also empty.
If  the book is full but rates are constantly under 0.05% - don't see the point in making a maxDailyRate in this scenario - will just cancel  and renew it endlessly.

**Comment by praeluceo** [comment]
I am also unable to use the bot as it hangs at "provided" see the error below:
$ python lendingbot.py 
2016-03-08 21:14:24 Welcome to Poloniex Lending Bot
2016-03-08 21:14:24 ERROR: 'provided'
2016-03-08 21:15:25 ERROR: 'provided'
2016-03-08 21:16:25 ERROR: 'provided'

My configuration looks as so (and I believe my API key & secret are correct):
`[BOT]
#sleep between iterations, time in seconds
sleeptime = 60
#minimum daily lend rate in percent
mindailyrate = 0.1
#max rate. 2% is good choice because it's default at margin trader interface. 5% is max to be accepted by the exchange
maxdailyrate = 4.9
#The number of offers to split the available balance uniformly across the [gaptop, gapbottom] range.
spreadlend = 3
#The depth of lendbook (in percent of lendable balance) to move through before placing the first (gapbottom) and last (gaptop) offer.
#if gapbottom is set to 0, the first offer will be at the lowest possible rate. However some low value is recommended (say 10%) to skip dust offers
gapbottom = 15
gaptop = 100
#Daily lend rate threshold after which we offer lends for 60 days as opposed to 2. If set to 0 all offers will be placed for a 2 day period
sixtydaythreshold = 0.2
# AutoRenew - if set to 1 the bot will toggle the AutoRenew flag for the loans when you stop it (Ctrl+C) and clear the AutoRenew flag when started
autorenew = 0
#custom config per coin, useful when closing positions etc.
#syntax: [COIN:mindailyrate:maxactiveamount, ... COIN:mindailyrate:maxactiveamount]
#if maxactive amount is 0 - stop lending this coin. in the future you'll be able to limit amount to be lent.
#coinconfig = ["BTC:0.18:1","CLAM:0.6:1"]
`

I've played with different configuration settings for autorenew and the min/max dailyrate variables, but nothing manages to get the bot running. And yes, there is a balance in the BTC Lending wallet.

**Comment by rreceptai** [comment]
Thank you for explaining, Mikadily and thanks for a great bot.

So the gapbottom and gaptop is the percentage amount of that snapshot of first loan offers(without loading the whole supply(load more)) right?

**Comment by rnevet** [comment]
The "load more" on the web page works in a different way,
The page actually loads all offers using a limit of 99999, initially displays only 50 and the "load more" display +100 more offers.

The bot doesn't supply a limit param - I'm not sure on what base does the server decide how many offers to return. 

The gaptop/gapbottom works relative to the snapshot of returned offers from the server.

**Comment by pmknutsen** [comment]
@Mikadily The default behavior (lending at max rate) is confusing. It makes more sense that if order book contains offers below my min lending offer, then the bot should simply place offers between my min and max offers. Then, there is still a chance they get picked up at that rate if someone comes along and vacuums current offerings. Now, if someone does that, and other bots place new offers again below my min rate I will never place offers at my min bid, and my funds will never be borrowed.

I think most bot users want their funds borrowed, even if that means at their min offer rates.

**Comment by Mikadily** [comment]
@pmknutsen the bot's purpose is to maximize the profits. Read my previous comment. The difference between lending rates in that case is massive 0.05% to 2% (or whatever your max) and it actually happened to me. Where as in your example it will be a 0.1-0.2% (rough guess).

But actually what you're asking is fairly easy to implement as a feature. How should I call configuration option?

**Comment by rnevet** [comment]
I think this is a very rare scenario, the actual situation with ETH is the orderbook is full but just not seen by the bot completely - otherwise it would set the offer to minDailyRate and not maxDailyRate.

**Comment by sanitariu** [comment]
minRate must be used exactly by name. Sometimes bot will lend at lower let's say 0.11 percent although you put 0.2 percent as minRate. And actually after 1-2 hours you can see 0.3 percent on the lending market. Now you gave it at 0.11 and you must wait 2 days. May be separate option for this to be mandatory minRate ?

**Comment by Mikadily** [comment]
> Sometimes bot will lend at lower let's say 0.11 percent although you put 0.2 percent as minRate.

@sanitariu that's impossible.

**Comment by Webonaute** [comment]
would be nice to keep track of average per day than have an option to not lend under like 10% of daily average. so the application will lend the money only at the best moment of the day. I think this can prevent ppl lending at ridiculus % once a day and lock lot of ppl for 2 day at very low rate and the rest of the day the % is like x2 or x4 of what you are lock to. not sure if what I say seem clear.

**Comment by dutu** [comment]
@Webonaute, average per day can be calculated, however when you stop lending if rate falls under the average, you're basically also stopping to carry on with average calculation. 

Secondly, yes your loan could be locked for 2 days at a very low rate. However, at the moment (for BTC) loan average-holding-time is 3h53m, as such probability to have your loan locked for 2 days is quite small at the moment. (ATM, for ETH alht is 1h14m).

But your questions is very valid: what is the rate for which it is better to wait instead of lending, and for how long do you wait until you decide is more profitable to lend. The answer to these question can be calculated/estimated, but you need a few other parameters (including the average loan-holding-time), which the current bot is not giving.

**Comment by dutu** [comment]
Average loan-holding-time is now 10h5m and increasing (due to low volatility I guess). Estimated waiting time for higher/more profitable rates is now much longer and increasing. So you see how this parameter impacts on estimating best profitability

**Comment by rnevet** [comment]
@dutu Where are you getting the loan-holding-time from?

**Comment by dutu** [comment]
I calculate it with tiny loans placed every minute

**Comment by robby-dermody** [comment]
Agree with @pmknutsen and @rnevet above.

I was running into this issue with several markets. Adding "&limit=1000" to the returnLoanOrders API request fixed it (i.e. by "fixed", I mean that now loans are made off of my configured minimum, and not just placing them at the max). The current behavior is a bit odd in that it a) only grabs a fraction of the book and b) lends at the max rate by default. This seems to lead to unpredictability (i.e. sometimes the rates were set off of my minimums, and sometimes off of the max rate, depending on the book composition at the time of run)

**Comment by Mikadily** [comment]
@robby-dermody I agree it's confusing, but only if you don't know what's happening under the hood.
We will have the "&limit=1000" thing as a feature soon.

**Comment by m33Lucky** [comment]
So how do I edit my coinconfig for Xmr , I want to start lending at 0.018 onwards , sorry new to this .thanks for all replys

**Comment by rnevet** [comment]
Should be solved with commit: a81cbbd "Dynamically increase limit of returnLoanOrders request"
